### PR TITLE
[lib] Try to handle critical exceptions from memprof.

### LINKED ISSUE #17760: coq is not robust to async interruptions (typically memprof limits)
The check `Global.assert_not_synterp` relies on a flag that is not part of the summary.

Thus, it is not possible to recover safely from async interruptions. The flag setting should be protected with `Fun.finally` or equivalent.

### PR DESCRIPTION
**Pull request by ejgallego**
Fixes #17760 ?

This is the last blocker for
https://github.com/ejgallego/coq-lsp/pull/509 to go ahead.

This has the potential to be quite costly tho, maybe we should explore some other alternatives?

Help with this is much welcome.

cc: @gadmm @Skyskimmer

### CODE REVIEW

**Comment by ejgallego** [comment]
> Can you not `@` in the commit message? It spams notifications on every amend/rebase.

Fixed sorry.

> Why is that?

The implementation of `is_interrupted` does quite a few more things than we are used to, it will depend on how much we call `noncritical`.

From a very quick peek at the stdlib, it seems quite random, but in some cases we call it really really a lot (I guess you know where that is ;))
```
...
noncritical called 9509 times              
noncritical called 3470 times              
noncritical called 9912 times              
noncritical called 13562 times             
noncritical called 77972 times             
noncritical called 23768 times             
noncritical called 7983 times              
noncritical called 1 times                 
noncritical called 273924 times            
noncritical called 273842 times            
noncritical called 2806 times              
noncritical called 2806 times              
noncritical called 15262 times             
noncritical called 7464 times              
noncritical called 1435101 times           
noncritical called 42757 times             
noncritical called 1 times                 
noncritical called 1 times                  
noncritical called 1 times          
....
```

We need to bench with the package installed, maybe the easiest is to update coq.opam?

**Comment by SkySkimmer** [comment]
>I guess you know where that is

I don't

**Comment by ejgallego** [comment]
My hypothesis for such a high number of noncritical calls was conversion and the catch-all games, but maybe it is indeed a different thing.

**Comment by gadmm** [comment]
You are right to wonder about the cost of `Memprof.is_interrupted`. If feedback is that it is too costly, then I could work to make it faster (ideally the cost of a TLS lookup).

**Comment by ejgallego** [comment]
> You are right to wonder about the cost of `Memprof.is_interrupted`. If feedback is that it is too costly, then I could work to make it faster (ideally the cost of a TLS lookup).

That could be very helpful indeed; let's run some benchs first. So far my understanding of the situation for Coq is:
- Coq relies heavily on `noncritical exn` , to the order of 10s of millions calls in some files (that we are really catching that many exns is a different issue, I'd suggest actually making a small patch to trace this)
- For memprof-limits to work, we need to recognize when we are in an exception context, and have Coq avoiding to catch those, otherwise we risk problems. The only supported way is using `is_interrupted`. Is my understanding correct @gadmm ?
- If the two previous points are correct, we have two choices:
  + sped up is_interrupted quite a lot, so we can use it in critical paths
  + implement two versions of `noncritical`, one that is resource-aware (and thus needs to call `is_interrupted`) and one that is not and thus cannot be used in context where resources have been taken.

**Comment by SkySkimmer** [comment]
I think we need to do the benching before coming up with theories.

**Comment by ejgallego** [comment]
Some data w.r.t. to exn general catch-all:
```
 1769: Raised at Ltac_plugin__Tactic_matching.verify_metas_coherence.merge in file "plugins/ltac/tactic_matching.ml", line 103, characters 13-37
Called from Stdlib__Map.Make.merge in file "map.ml", line 395, characters 44-63
Called from Stdlib__Map.Make.merge in file "map.ml", line 395, characters 64-79
Called from Stdlib__Map.Make.merge in file "map.ml", line 395, characters 64-79
Called from Ltac_plugin__Tactic_matching.verify_metas_coherence in file "plugins/ltac/tactic_matching.ml", line 108, characters 11-36
Called from Ltac_plugin__Tactic_matching.PatternMatching.merge in file "plugins/ltac/tactic_matching.ml", line 166, characters 14-42

 2849: Raised at Loc.raise in file "lib/loc.ml", line 101, characters 16-23
Called from Unification.unify_0_with_initial_metas in file "pretyping/unification.ml", line 1244, characters 13-48
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Unification.w_unify_core_0 in file "pretyping/unification.ml", line 1618, characters 5-95
Called from Unification.make_pattern_test.matching_fun in file "pretyping/unification.ml", line 1728, characters 18-87

 3243: Raised at Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 480, characters 76-104
Called from CArray.fold_left2.fold in file "clib/cArray.ml", line 269, characters 33-62
Called from Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 359, characters 14-84
Called from Constr_matching.extended_matches in file "pretyping/constr_matching.ml", line 492, characters 25-58
Called from Ltac_plugin__Tactic_matching.PatternMatching.pattern_match_term in file "plugins/ltac/tactic_matching.ml", line 243, characters 22-77

19456: Raised at Int.Map.find in file "clib/int.ml", line 40, characters 14-29
Called from Dyn.Make.Self.Map.find in file "clib/dyn.ml", line 130, characters 26-46
Called from Store.Make.get in file "clib/store.ml", line 46, characters 13-27

30262: Raised at Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 480, characters 76-104
Called from Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 359, characters 47-74
Called from Constr_matching.extended_matches in file "pretyping/constr_matching.ml", line 492, characters 25-58
Called from Ltac_plugin__Tactic_matching.PatternMatching.pattern_match_term in file "plugins/ltac/tactic_matching.ml", line 243, characters 22-77

52666: Raised at Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 480, characters 76-104
Called from Constr_matching.extended_matches in file "pretyping/constr_matching.ml", line 492, characters 25-58
Called from Ltac_plugin__Tactic_matching.PatternMatching.pattern_match_term in file "plugins/ltac/tactic_matching.ml", line 243, characters 22-77
[./theories/Structures/GenericMinMax.v] noncritical called 126078 times
```

**Comment by ejgallego** [comment]
> I think we need to do the benching before coming up with theories.

Yes I'm updating the PR so memprof-limits is installed in the bench; but I think it is fine to predict some slowdown based on the data we already have.

**Comment by ejgallego** [comment]
@coqbot: bench

**Comment by gadmm** [comment]
Yes, `is_interrupted` is the designated mechanism. You cannot know if the user code wraps the exception in another one, etc.

I prefer if Coq uses `is_interrupted` does not make its code more complicated, even if I have to work to optimise it.

**Comment by SkySkimmer** [comment]
Doesn't seem like much change.

What about if you run with some resource limit active?

**Comment by gadmm** [comment]
We saw a bigger impact of memprof-limits with @SkySkimmer's experiments. Is memprof running at any point with this PR or does this require actual interaction from the user?

**Comment by ejgallego** [comment]
> We saw a bigger impact of memprof-limits with @SkySkimmer's experiments. Is memprof running at any point with this PR or does this require actual interaction from the user?

Only thing we are doing here is to call `is_interrupted ()`, memprofs-limits is not even initialized in the bench.

But indeed I have measured the impact with the `fcc` Flèche-based compiler and memprof-limits active, and I couldn't spot any impact. What `fcc` does is: a) setups the mp runtime, b) creates a token that will never be set; c) single-thread; d) calls Coq 2/3 times per-sentence to parse and compile the document.

Under that scenario, I didn't measure any time impact. I hacked the code to also add the global limit and the allocation limit; performance was also not impacted.

I abuse this thread to ask a related question @gadmm : is the `Thread_map` library compatible with OCaml 5.x ?

I think that is a fantastic library and Coq could really take advantage of having good thread-local storage. For example that would allow coq-lsp to have a reader thread that would use its own, consistent summary, while having some other, serialized, write-commands.

**Comment by ejgallego** [comment]
I got scared about performance in exception-related code as if you run `coqc` with `-bt`, that will impact performance 20-30% on average; that's a lot. And indeed, I see we are hitting code paths with ` | exn when noncritical exn -> ...` a lot more of what is IMHO desirable.

Catch-all clauses should only be used at the barriers between components, for example.

**Comment by ejgallego** [comment]
I did some more performance research, as far as I can see the differences are very small, and thus quite hard to measure.

IMHO it is working very well, we could indeed the support for Coq if we want, I will support `memprof-limits` in the next coq-lsp release. For the users that need it, it is a very good feature at little cost.

**Comment by gadmm** [comment]
re Thread_map (assuming you fix Mutex_aux for OCaml 5 and remove the dependency on Masking), good question. It might be correct for multicore (notice the innocuous data race regarding the mutable field).

But better use something like https://github.com/c-cube/thread-local-storage which is going to be much more efficient anyway (I still need to audit it for robustness against async exceptions).

**Comment by gadmm** [comment]
Regarding your perf experiments, this is good news. To confirm the good news, could you please do some control experiment, to make sure that the right thing is measured? I propose to re-do the test with a sampling rate which you know for sure that it should affect the performance. You can change the sampling rate by running a dummy profiling session using `Memprof_limits.Memprof.start` with chosen sampling_rate (e.g. 1e-2, which is very large), callstack_size of 0, and the null_tracker.

**Comment by ejgallego** [comment]
Good idea; it took me a while to realize how to use `Memprof_limits.Memprof.start`, in particular I did not realize I should still call the main `Memprof_limits.start_memprof_limits` still after.

I can confirm that setting the values to `1e-2` or `1e-1` considerably slows down the program. So indeed, that testing seems valid in that sense.

**Comment by gadmm** [comment]
thanks!

**Comment by ejgallego** [comment]
Thanks to you! Current experiments can be reproduced with the tree at https://github.com/ejgallego/coq-lsp/pull/509

In particular:
```
$ make && dune exec -- /usr/bin/time fcc --int_backend=Coq --display=verbose --root examples examples/Rtrigo1.v
```
for example, change `--int_backend=Mp` and you will use the `memprof-limits` backend.

To slow the program down, edit the file `coq/limits_mp_impl.real.ml` and set `oc_mode` to `true`. Re-run.

**Comment by SkySkimmer** [comment]
So what was it that had
>noncritical called 1435101 times           

?

**Comment by ejgallego** [comment]
> So what was it that had
> 
> > noncritical called 1435101 times
> 
> ?

@SkySkimmer that's:

```
[1435101] noncritical calls in ./theories/MSets/MSetRBT.v
```

there is also:
```
[1300483] noncritical calls in ./theories/FSets/FMapAVL.v
```

Indeed, good idea, I did test this file with `fcc` and all limits on + this PR. I couldn't spot a difference, tho I don't have the right machinery to do precision benchmarking. I also compared to plain `master`, timings on `MSetRBT` seem similar.

I guess that means this PR works fine as-is, pending of course a future optimization of the interrupt lookup.

But it seems that as of today, the few cycles added don't amount to much measurable time.

**Comment by ejgallego** [comment]
Some interesting stuff:

```
[  17357] noncritical calls in ./theories/Classes/RelationPairs.v
16953: <in exception printer>: Anomaly "Uncaught exception Not_found."
Please report at http://coq.inria.fr/bugs/.
<original exception>:
Uncaught exception Pretype_errors.PretypeError(_, _, _).
```

```
[   4999] noncritical calls in ./theories/Structures/OrdersAlt.v
45597: This expression should be a simple identifier.
```

```
...
ncalls: 33281: <in exception printer>: Anomaly "Uncaught exception Not_found."
Please report at http://coq.inria.fr/bugs/.
Raised at Int.Map.find in file "clib/int.ml", line 40, characters 14-29
Called from HMap.Make.find in file "clib/hMap.ml", line 364, characters 12-28
Called from Nametab.shortest_qualid_of_global in file "library/nametab.ml", line 530, characters 18-75
Called from Constrextern.default_extern_reference in file "interp/constrextern.ml", line 200, characters 6-51
Called from Constrextern.extern_reference in file "interp/constrextern.ml" (inlined), line 208, characters 35-62
Called from Constrextern.extern in file "interp/constrextern.ml", line 1001, characters 22-58
Called from Flags.with_options in file "lib/flags.ml", line 37, characters 12-15
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Himsg.display_eq in file "vernac/himsg.ml", line 132, characters 12-78
Called from Himsg.pr_explicit_aux in file "vernac/himsg.ml", line 143, characters 14-47
Called from Himsg.pr_explicit in file "vernac/himsg.ml", line 177, characters 15-61
Called from Himsg.explain_cannot_unify in file "vernac/himsg.ml", line 740, characters 15-40
Called from Himsg.wrap_unhandled in file "vernac/himsg.ml", line 1575, characters 11-16
Called from CErrors.print_gen in file "lib/cErrors.ml", line 108, characters 10-13
Called from CErrors.print_gen in file "lib/cErrors.ml", line 119, characters 17-51

<original exception>:
Uncaught exception Pretype_errors.PretypeError(_, _, _).
Raised at Loc.raise in file "lib/loc.ml", line 101, characters 16-23
Called from Unification.unify_0_with_initial_metas in file "pretyping/unification.ml", line 1248, characters 14-64
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Unification.w_unify_core_0 in file "pretyping/unification.ml", line 1618, characters 5-95
Called from Clenv.unify.(fun) in file "proofs/clenv.ml", line 966, characters 17-49

Raised at Loc.raise in file "lib/loc.ml", line 101, characters 16-23
Called from Unification.unify_0_with_initial_metas in file "pretyping/unification.ml", line 1248, characters 14-64
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Unification.w_unify_core_0 in file "pretyping/unification.ml", line 1618, characters 5-95
Called from Clenv.unify.(fun) in file "proofs/clenv.ml", line 966, characters 17-49

ncalls: 110377: No matching clauses for match.
Raised at Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 480, characters 76-104
Called from Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 359, characters 47-74
Called from Constr_matching.extended_matches in file "pretyping/constr_matching.ml", line 492, characters 25-58
Called from Ltac_plugin__Tactic_matching.PatternMatching.pattern_match_term in file "plugins/ltac/tactic_matching.ml", line 243, characters 22-77

Raised at Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 480, characters 76-104
Called from Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 359, characters 47-74
Called from Constr_matching.extended_matches in file "pretyping/constr_matching.ml", line 492, characters 25-58
Called from Ltac_plugin__Tactic_matching.PatternMatching.pattern_match_term in file "plugins/ltac/tactic_matching.ml", line 243, characters 22-77

ncalls: 246173: Not convertible.
Raised at Loc.raise in file "lib/loc.ml", line 101, characters 16-23
Called from Proofview.Goal.enter.f in file "engine/proofview.ml", line 1134, characters 40-46

Raised at Loc.raise in file "lib/loc.ml", line 101, characters 16-23
Called from Proofview.Goal.enter.f in file "engine/proofview.ml", line 1134, characters 40-46

ncalls: 334742: Tactic failure.
Raised at Int.Map.find in file "clib/int.ml", line 40, characters 14-29
Called from Dyn.Make.Self.Map.find in file "clib/dyn.ml", line 130, characters 26-46
Called from Store.Make.get in file "clib/store.ml", line 46, characters 13-27

Raised at Int.Map.find in file "clib/int.ml", line 40, characters 14-29
Called from Dyn.Make.Self.Map.find in file "clib/dyn.ml", line 130, characters 26-46
Called from Store.Make.get in file "clib/store.ml", line 46, characters 13-27

ncalls: 523282: No matching clauses for match.
Raised at Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 480, characters 76-104
Called from Constr_matching.extended_matches in file "pretyping/constr_matching.ml", line 492, characters 25-58
Called from Ltac_plugin__Tactic_matching.PatternMatching.pattern_match_term in file "plugins/ltac/tactic_matching.ml", line 243, characters 22-77

Raised at Constr_matching.matches_core.sorec in file "pretyping/constr_matching.ml", line 480, characters 76-104
Called from Constr_matching.extended_matches in file "pretyping/constr_matching.ml", line 492, characters 25-58
Called from Ltac_plugin__Tactic_matching.PatternMatching.pattern_match_term in file "plugins/ltac/tactic_matching.ml", line 243, characters 22-77

[1435101] noncritical calls in ./MSetRBT.v
```

**Comment by SkySkimmer** [comment]
Quick hack to add noncritical call counts to the profiler
~~~diff
modified   lib/cErrors.ml
@@ -149,8 +149,10 @@ let _ = register_handler begin function
     by inner functions during a [vernacinterp]. They should be handled
     only at the very end of interp, to be displayed to the user. *)
 
+let nc_count = ref 0
+
 [@@@ocaml.warning "-52"]
-let noncritical = function
+let noncritical e = incr nc_count; match e with
   | Sys.Break | Out_of_memory | Stack_overflow
   | Assert_failure _ | Match_failure _ | Anomaly _
   | Control.Timeout -> false
modified   lib/cErrors.mli
@@ -77,3 +77,5 @@ val noncritical : exn -> bool
 val register_additional_error_info
   :  (Exninfo.info -> Pp.t option)
   -> unit
+
+val nc_count : int ref
modified   lib/newProfile.ml
@@ -154,6 +154,9 @@ let make_mem_diff ~(mstart:Gc.stat) ~(mend:Gc.stat) =
   [("major_words",ppf major_words); ("minor_words", ppf minor_words);
    ("major_collect", ppi major_collect); ("minor_collect", ppi minor_collect); ]
 
+let make_nc_diff ~ncstart ~ncend =
+  [("noncritical", `Intlit (string_of_int (ncend - ncstart)))]
+
 let make_instr_diff ~istart ~iend =
   match System.instructions_between ~c_start:istart ~c_end:iend with
   | Ok count -> [("instr", `Intlit (Int64.to_string count))]
@@ -176,23 +179,28 @@ let profile name ?args f () =
     enter name ?args ();
     let mstart = Gc.quick_stat () in
     let istart = Instr.read_counter () in
+    let ncstart = !CErrors.nc_count in
     let v = try f ()
       with e ->
         let e = Exninfo.capture e in
         let iend = Instr.read_counter () in
         let mend = Gc.quick_stat () in
+        let ncend = !CErrors.nc_count in
         let args =
           make_instr_diff ~istart ~iend @
-          make_mem_diff ~mstart ~mend
+          make_mem_diff ~mstart ~mend @
+          make_nc_diff ~ncstart ~ncend
         in
         leave name ~args ();
         Exninfo.iraise e
     in
     let iend = Instr.read_counter () in
     let mend = Gc.quick_stat () in
+    let ncend = !CErrors.nc_count in
     let args =
       make_instr_diff ~istart ~iend @
-      make_mem_diff ~mstart ~mend
+      make_mem_diff ~mstart ~mend @
+      make_nc_diff ~ncstart ~ncend
     in
     leave name ~args ();
     v
@@ -241,9 +249,11 @@ let finish () = match !accu with
   | Some { ch } ->
     let iend = Instr.read_counter () in
     let mend = Gc.quick_stat () in
+    let ncend = !CErrors.nc_count in
     let args =
       make_instr_diff ~istart:global_start_count ~iend @
-      make_mem_diff ~mstart:global_start_stat ~mend
+      make_mem_diff ~mstart:global_start_stat ~mend @
+      make_nc_diff ~ncstart:0 ~ncend
     in
     leave "process" ~last:"" ~args ();
     Format.fprintf ch "],\n\"displayTimeUnit\": \"us\" }";
~~~
Result on MSetRBT
[MSetRBT.json.gz](https://github.com/coq/coq/files/14923697/MSetRBT.json.gz)
Not sure there's much interesting to get from it.

Looks like 1269713 (vast majority) of the calls are the check in Proofview.tclZERO.

**Comment by SkySkimmer** [comment]
>Also looks like 1269713 (vast majority) of the calls are the check in Proofview.tclZERO (I did a quick test by removing that check).

and of the remaining 165388, 131932 are from Proofview.catchable_exception (used mostly in `enter`)

**Comment by ejgallego** [comment]
> Also looks like 1269713 (vast majority) of the calls are the check in Proofview.tclZERO (I did a quick test by removing that check

> 131932 are from Proofview.catchable_exception (used mostly in `enter`)

I see, thanks a lot for the data. The only way to reduce this kind of call (would we like to) is I guess to forbid tactic code from raising exceptions (instead requiring it to use `tclZERO` directly)

I found this issue before somehow ? cc: #15937

**Comment by SkySkimmer** [comment]
??
Using tclZERO directly won't reduce use of tclZERO though?

**Comment by ejgallego** [comment]
> ?? Using tclZERO directly won't reduce use of tclZERO though?

At some point we could have a "fast-path" that wouldn't call `noncritical`, but indeed, that's orthogonal, I think I went a bit offtopic with this; it seems this patch is OK, indeed enabling limitis could have some 1-2% impact, I couldn't measure it relialby.

**Comment by SkySkimmer** [comment]
Do we need to declare memprof limits as an optional dependency in dune-project / coq-core.opam?

File: lib/noncritical.mli [hunk 1]
```diff
@@ -0,0 +1,2 @@
+exception Anomaly of string option * Pp.t (* System errors *)
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply lib/noncritical.mli:1]: sounds Ok to me

**Comment by ejgallego** [comment]
Should look better now / be ready.

**Comment by ejgallego** [comment]
Likely overkill but let's do a full CI to test OCaml versions etc...

@coqbot run full CI

File: lib/interrupted.mli [hunk 2]
```diff
@@ -0,0 +1 @@
+val is_interrupted : unit -> bool [@@inline]
```
Comment (SkySkimmer) [inline lib/interrupted.mli:1]: I thought inline in mli does nothing?

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
coq-core.opam needs  updating

**Comment by ejgallego** [comment]
> coq-core.opam needs updating

Oh damn, I had indeed updated it, but I think I did update it in the wrong place; will fix ASAP.

**Comment by ejgallego** [comment]
@coqbot run full CI

**Comment by ejgallego** [comment]
At some point we may have a CI choice to make if projects start depending on `memprof-limits`, for now, `coq-lsp` has only an optional dependency.

File: lib/interrupted.mli (hunk 2 above)
Comment (SkySkimmer) [reply lib/interrupted.mli:1]: ping

Comment (ejgallego) [reply lib/interrupted.mli:1]: oh yes :facepalm: one sec

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Comment by ejgallego** [comment]
#19133 means that we may want to update the milestone here.

**Pull request merged**